### ISSUE TITLE
Release 0.5.0

### ISSUE THREAD

**Issue by nchashch** [amethyst/amethyst#241]
This is the issue for tracking the progress of 0.5.0 release and for coordinating the release (what should be merged, what should be deferred until 0.6.0, in what order PRs should be merged).

I think we should release 0.5.0 when each of the current open PRs is either merged or deferred until 0.6.0. Also, I think any new PRs should go into 0.5.X releases (non-breaking ones) or into 0.6.0 release (major breaking ones), so that we have a finite amount of work to do before 0.5.0 release.
# Currently open PRs
When a PR is reviewed, approved, and merged it should be ticked off. And if a PR is deferred until 0.6.0 it should also be ticked off (marked as [deferred]).
## Major breaking changes:
- [x] #265 - Audio support
- [x] #233 - Renderer rewrite
- [x] #221 - Breaking change of `State` methods' arguments + refactor of the main loop iteration [closed]
- ~~#195 - Asset manager rewrite~~
- [x] #244 - Asset manager rewrite
- [x] #152 - Configuration mechanism rewrite

## Minor breaking changes:
- [x] #186 - Add mouse movement handling to `InputHandler`
- [x] #247 - Input rebind patch
- [x] #250 - Convert to specs v0.9
- [x] #306 - Restructure the source tree for more functionality driven design

## Non-breaking changes:
- [x] #232 - Add pong tutorial [deferred]

## TODO:
- [x] Publish 0.5.0 on crates.io
- [x] Update CHANGELOG.md
- [x] Write and publish another These Months in Amethyst post

**Comment by derekdreery** [comment]
I'd defer the pong tutorial, as I need to see how the renderer has changed before I can do that, and it shouldn't block release.

**Comment by nchashch** [comment]
@derekdreery Ok. Updating the issue.

**Comment by nchashch** [comment]
I think we can merge #186, #221, and #152 (as soon as it passed CI and was reviewed). This would introduce breaking changes in `develop` branch so there won't be any PATCH releases until 0.5.0 for sure. After these PRs were merged @ebkalderon and @torkleyy can rebase their WIP PRs on top of `develop` and resolve merge conflicts. Then only finishing WIP PRs and resolving merge conflicts between #195 and #233 would be left before 0.5.0 release.

**Comment by nchashch** [comment]
#152 will be merged after @torkleyy 's review.

**Comment by nchashch** [comment]
All PRs except #233 and #195 are merged, deferred, or closed. So once they are completed, reviewed, and merged we can release 0.5.0.

**Comment by torkleyy** [comment]
I'm essentially finished with the asset management, just need to write documentation and implement the formats we had, but that shouldn't take me long. I don't know however, how I should include it into the engine. More specifically, I'm not sure if it should go into a separate crate (on my account, amethyst, or in-tree) or just be a module. Given that I think it might be useful for others, I somewhat tend to the former options.

**Comment by nchashch** [comment]
Added #247 to minor breaking changes.

**Comment by nchashch** [comment]
Added #250 to minor breaking changes.

**Comment by ebkalderon** [comment]
@/all I think we should plan future releases using [GitHub Projects](https://github.com/amethyst/amethyst/projects). What if I removed the three existing projects and we instead create a *Release 0.5* project, then a *Release 0.6* project, and so on? That way we won't need to use the issue tracker for this.

Still, let's keep this issue open for now until 0.5 lands, at least.

**Comment by Xaeroxe** [comment]
@ebkalderon 👍 I like that plan.

**Comment by Xaeroxe** [comment]
I'm fairly certain I can squeeze audio in in time for the 0.5.0 release.  It's already most of the way there, I just need to implement better music controls.

**Comment by nchashch** [comment]
So, we will release 0.5.0 after #265 is reviewed and merged. All new features would then go either into 0.5.X releases or into 0.6.0.

**Comment by Xaeroxe** [comment]
Changelog has been kept up to date as progress has been ongoing, so I checked off the CHANGELOG.md box.

**Comment by Xaeroxe** [comment]
So in order to avoid any more feature creep for this release I've added a new branch called release-0.5.0. New PRs should go to our develop branch which will be for 0.6.0, and I've switched the PRs we've designated for 0.5.0 to merge into release-0.5.0. Any objections?

**Comment by torkleyy** [comment]
We should explicitly say that the asset manager is still going to change in some places.

**Comment by nchashch** [comment]
@torkleyy You mean before 0.5.0 release or after?

**Comment by torkleyy** [comment]
I mean we should say that we release a new version, but that the asset stuff will certainly have breaking changes in the next release.

**Comment by nchashch** [comment]
I see. I think we can add it into the TMiA post, write that in 0.6.0 release asset management will have breaking changes.

**Comment by Xaeroxe** [comment]
Same for the AudioSystem.  We've already found a corner case where it doesn't work as well.

**Comment by nchashch** [comment]
Actually, I think it would be better to include these changes into 0.5.X releases since they are changes in already existing modules and this way they can be rolled out quicker because they wouldn't be packaged up with all the other things in 0.6.0.

It would break the "no non-breaking changes in PATCH" releases rule, but I don't think anybody cares about this rule because amethyst is explicitly unstable. To accommodate breaking tweaks to existing code, it might make sense to drop any "next release won't break the API" guarantees.

So PATCH releases would just mean small changes and MINOR releases would just mean big changes, without any kind of stability guarantee. AFAIK semver doesn't specify how releases should behave before 1.0.0, so it would be ok. Although it would be nice to state this explicitly in TMiA.

**Comment by torkleyy** [comment]
I don't think breaking semver guarantees makes any sense here. There's no harm in releasing new versions more often.

**Comment by nchashch** [comment]
@torkleyy Yes, you are right. Small -- big distinction is less useful for users than breaking -- non-breaking distinction. I guess we will have to enforce this, so that contributors mark their PRs as either breaking or non-breaking and target them at a PATCH or a MINOR release.

Then bug-fixes, refactors, internal optimizations, documentation, comments would go into PATCH releases and new features and API changes would go into MINOR releases.

**Comment by Xaeroxe** [comment]
Please merge https://github.com/amethyst/amethyst/pull/307 before releasing.  It's a couple small changes that make the InputHandler and AudioSystem much easier to use correctly.

**Comment by Xaeroxe** [comment]
I will perform the release, before I do I'd like an explicit sign-off from at least two other organization members.  Pinging @amethyst/engine-devs

**Comment by Xaeroxe** [comment]
Since no one was willing to say yes or no I went ahead and did it.  0.5.0 is published!

**Comment by kvark** [comment]
@Xaeroxe thank you!
would be great to spread the news around, optimally with "Write and publish another These Months in Amethyst post" task accomplished

**Comment by Xaeroxe** [comment]
I agree!

**Comment by Xaeroxe** [comment]
Although before we draw too much attention to this release it'd be nice to have a good runway for new users, to that end I'd like to rewrite the book.  I'll get started on that.

**Comment by Rhuagh** [comment]
So I guess release-0.5.0 becomes master when we're ready for attention?

**Comment by Xaeroxe** [comment]
Well I failed in my plan to rewrite the book, got distracted by other tasks.  However everything else here is done, so I'm going to close this.
